runtime/derive/src/lib.rs

```diff
@@ -0,0 +1,176 @@
+        Fields::Named(fields) => fields
+            .named
+            .iter()
+            .map(|field| field.ident.as_ref().unwrap())
+            .collect(),
+        _ => panic!("AscType can only be derived for structs with named fields"),
+    };
+    let field_names2 = field_names.clone();
+    let field_names3 = field_names.clone();
+    let field_types = match &item_struct.fields {
+        Fields::Named(fields) => fields.named.iter().map(|field| &field.ty),
+        _ => panic!("AscType can only be derived for structs with named fields"),
+    };
+
+    TokenStream::from(quote! {
+        impl#impl_generics AscType for #struct_name#ty_generics #where_clause {
+            fn to_asc_bytes(&self) -> Vec<u8> {
+               let mut bytes = Vec::new();
+                #(bytes.extend_from_slice(&self.#field_names.to_asc_bytes());)*
+
+                // Assert that the struct has no padding.
+                assert_eq!(bytes.len(), size_of::<Self>());
+                bytes
+            }
+
+            #[allow(unused_variables)]
+            fn from_asc_bytes(asc_obj: &[u8]) -> Self {
+                assert_eq!(asc_obj.len(), size_of::<Self>());
+                let mut offset = 0;
+
+                #(
+                let field_size = std::mem::size_of::<#field_types>();
+                let #field_names2 = AscType::from_asc_bytes(&asc_obj[offset..(offset + field_size)]);
+                offset += field_size;
+                )*
+
+                Self {
+                    #(#field_names3,)*
+                }
+            }
+        }
+    })
+}
+
+// Example input:
+// #[repr(u32)]
+// #[derive(AscType)]
+// enum JsonValueKind {
+//     Null,
+//     Bool,
+//     Number,
+//     String,
+//     Array,
+//     Object,
+// }
+//
+// Example output:
+// impl AscType for JsonValueKind {
+//     fn to_asc_bytes(&self) -> Vec<u8> {
+//         let discriminant: u32 = match *self {
+//             JsonValueKind::Null => 0u32,
+//             JsonValueKind::Bool => 1u32,
+//             JsonValueKind::Number => 2u32,
+//             JsonValueKind::String => 3u32,
+//             JsonValueKind::Array => 4u32,
+//             JsonValueKind::Object => 5u32,
+//         };
+//         discriminant.to_asc_bytes()
+//     }
+//
+//     fn from_asc_bytes(asc_obj: &[u8]) -> Self {
+//         let mut u32_bytes: [u8; size_of::<u32>()] = [0; size_of::<u32>()];
+//         u32_bytes.copy_from_slice(&asc_obj);
+//         let discr = u32::from_le_bytes(u32_bytes);
+//         match discr {
+//             0u32 => JsonValueKind::Null,
+//             1u32 => JsonValueKind::Bool,
+//             2u32 => JsonValueKind::Number,
+//             3u32 => JsonValueKind::String,
+//             4u32 => JsonValueKind::Array,
+//             5u32 => JsonValueKind::Object,
+//             _ => panic!("value {} is out of range for {}", discr, "JsonValueKind"),
+//         }
+//     }
+// }
+fn asc_type_derive_enum(item_enum: ItemEnum) -> TokenStream {
```

Comment:
What happens if one of the enum variants has data fields? Can that be checked with an assert?

Reply:
The assert won't check that, but the match won't make sense and will not compile.